### ISSUE TITLE
Crash: Optimizations - game object lookup (Hyper-V)

### ISSUE THREAD

**Issue by drake127** [nwnxee/unified#1297]
Versions:
- NWN: 8193.21 (77c0c8a138c8379eaa467886333b101b33c6be48)
- NWNX: build8193.21-HEAD (d4023e379f02db462b18c33f6a643d93e678e16e)
It is also reproducible with 8193.20.1 and matching NWNX.

When I have `export NWNX_OPTIMIZATIONS_GAME_OBJECT_LOOKUP=true` set, the server crashes on exit with segmentation fault. I narrowed it down to this option with other NWNX plugins and other optimizations disabled.

No prior error is observed in logs:
```
exit
Server: Exiting...
I [12:27:37] [NWNXLib] [Plugin.cpp:81] Unloading plugin NWNX_Optimizations
Segmentation fault
```

I have core dump available but without symbols, there's not much I can do about it: https://files.planescape.cz/tmp/nwnx/core.7z

Can somebody confirms this issue is reproducible outside my environment?

**Comment by WilliamDraco** [comment]
The NWNX_Core option:
`NWNX_CORE_HARD_EXIT` | 0-1| 0 | If set, NWNX will hard kill the process after it unloads.
was added in #1272 due to this issue occuring with WSL.

**Comment by drake127** [comment]
Hi, thanks for a pointer. Do you have any idea why is this happening? I am running NWN server under Hyper-V, not WSL. Is this some virtualization issue I want to be aware of?

It doesn't hang, it crashes, so I was expecting some kind of double-free or similar.

**Comment by drake127** [comment]
You are correct though, that it doesn't happen on QEMU VM. I wonder why it is.

**Comment by mtijanic** [comment]
For context, why this happens is that the optimization takes over a global NWN object (CGameObjectArray). This global object is deinitialized by the game _after_ NWNX is fully unloaded, and this triggers a use-after-free bug. The bug is not always triggered and when it is, the exact behavior varies depending on what was happening in game previously.

Unfortunately, since NWNX is unloaded, we can't actually prevent this code from running by e.g. making a hook. Some of the actual fix options here are:

- asm hotpatch the function with NOPs in a way that doesn't get reverted when NWNX unloads
- Move the NWNX shutdown later in the sequence so we can hook it
- Have the GOA lookup optimization create a dummy valid base game object for the game to destroy afterwards

But generally the benefit is marginal and HARD_EXIT works just fine. We might make HARD_EXIT the default and simplify a lot of the other unload code too.
